### PR TITLE
Raise exception when no objects are on image

### PR DESCRIPTION
**Pull request by g-braeunlich**
Closes: #285

### CODE REVIEW

**Comment by rmjarvis** [comment]
I don't think this is the right place to raise an exception.  It can be completely fine to have no objects land on a particular sensor.  The problem in Virginia's case was that there were no objects in the input catalog, but then she explicitly said to draw 500 objects.  If she let `nobjects` be computed automatically (i.e. use all the input objects), it would have come out as zero and it would have happily drawn all (i.e. none) of them and moved on.
So I think we want
a) A higher verbosity message that there are no objects found (nuse==0).  Either logger.error or logger.warning.
b) An error raised in getObj if there are no objects available, but the user nonetheless requests one.

**Comment by g-braeunlich** [comment]
> So I think we want a) A higher verbosity message that there are no objects found (nuse==0). Either logger.error or logger.warning. b) An error raised in getObj if there are no objects available, but the user nonetheless requests one.

This should be done now

File: imsim/skycat.py [hunk 1]
```diff
@@ -52,6 +52,9 @@ def __init__(self, file_name, wcs, bandpass, obj_types=None,
         region = skyCatalogs.Box(*get_radec_limits(wcs, logger, edge_pix)[:4])
         self.objects = sky_cat.get_objects_by_region(region,
                                                      obj_type_set=obj_types)
+        if not self.objects:
+            raise RuntimeError("No skyCatalogs objects on image")
```
Comment (rmjarvis) [inline imsim/skycat.py:56]: I don't think this should be an error.  I believe this is the correlate to the place in instcat where you warn about "No objects found on image".  Thus, I think this should also be just a logged warning.

**Review by rmjarvis — CHANGES_REQUESTED** [review]
This still needs two things:
1. Rebase to the current main to get the tests working.
2. Add unit tests that check the new functionality.  At least checking that the new error is raised in the appropriate situation.

File: imsim/skycat.py (hunk 1 above)
Comment (g-braeunlich) [reply imsim/skycat.py:56]: OK. But what should I return if the we really hit the invalid access of item 0 in the empty list?
Or how would you prevent reaching this piece of code?

Comment (g-braeunlich) [reply imsim/skycat.py:56]: What I also thought:
Cannot we interpret `image.nobjects` as an upper limit? Or is there a usecase where a user really wants to have exact that many objects?

Comment (rmjarvis) [reply imsim/skycat.py:56]: It's the same thing as you did in instcat, isn't it?  It should only be an error if the user explicitly asks for more objects than exist.  But at this point, it should be allowed for there to be 0 objects on the image.

Comment (g-braeunlich) [reply imsim/skycat.py]: OK. Rebased, added a quick test and turned the above exception into a warning

**Comment by rmjarvis** [comment]
Why isn't there a corresponding error raised in the skycat `getObj`?  Is that possibility already handled gracefully somehow?  I would have thought it would have the same issue.

Also, maybe this is just a philosophical difference on my part, but since this issue stemmed directly from user error, I would have liked to see a test that recapitulated what Virginia did to show that it now produces better messaging than previously.

**Comment by g-braeunlich** [comment]
> Why isn't there a corresponding error raised in the skycat `getObj`? Is that possibility already handled gracefully somehow? I would have thought it would have the same issue.

Actually we also have `skycat_obj = self.objects[index]` there. Therefore I also added a check there.

> Also, maybe this is just a philosophical difference on my part, but since this issue stemmed directly from user error, I would have liked to see a test that recapitulated what Virginia did to show that it now produces better messaging than previously.

Are you thinking of an integration test? I personally would try to keep tests simple (the more test code the more maintenance)  while of course covering the relevant piece of code.

File: tests/test_skycat.py [hunk 2]
```diff
@@ -69,15 +76,17 @@ def test_get_wl_params(self):
...
-            self.assertAlmostEqual(g2, gamma2/(1. - kappa))
-            self.assertAlmostEqual(mu, 1./((1. - kappa)**2 - (gamma1**2 + gamma2**2)))
+            gamma1 = row["shear_1"]
+            gamma2 = row["shear_2"]
+            kappa = row["convergence"]
+            self.assertAlmostEqual(g1, gamma1 / (1.0 - kappa))
+            self.assertAlmostEqual(g2, gamma2 / (1.0 - kappa))
+            self.assertAlmostEqual(
+                mu, 1.0 / ((1.0 - kappa) ** 2 - (gamma1**2 + gamma2**2))
+            )
```
Comment (rmjarvis) [inline tests/test_skycat.py:89]: You ran black on this file I guess...?  There are a lot of content-free changes here.

**Review by rmjarvis — APPROVED** [review]

File: tests/test_skycat.py (hunk 2 above)
Comment (g-braeunlich) [reply tests/test_skycat.py:89]: Yes, sorry. Forgot to turn my auto formatting off. I will remove the reformatting again

Comment (g-braeunlich) [reply tests/test_skycat.py]: Done

**Pull request merged**